### ISSUE TITLE
residue name and dihedral recognition

### ISSUE THREAD

**Issue by evanfeinberg** [mdtraj/mdtraj#805]
Dear MSMBuilders,

I've been attempting to featurize with phi, psi, and chi2 dihedral angles a set of trajectories, each of the same receptor in slightly different conditions. For example, in one simulation, a residue might be a charged Asp, while in another, it is neutral. However, the publishers of these simulations renamed the charged residue as "ASH". In one simulation, histidine is simulated in its delta-protonated configuration ("HID") while in another in its epsilon-protonated ("HIE). Therefore, when I tell msmbuilder to only read in protein atoms, with selection language:

top = md.load_frame(filename, 0).topology
atom_indices = [a.index for a in top.atoms if a.residue.chain.index == 0 and a.residue.is_protein]
traj = md.load(filename, stride=stride, atom_indices=atom_indices)

, it includes the Asp for one simulation but does not when it has been renamed to Ash.

The following circumvention does not work:
atom_indices = [a.index for a in top.atoms if a.residue.chain.index == 0 and a.residue != "POPC"]

I think it is because it is only calculating dihedrals for residue names it "recognizes" (e.g. HID), but for those it doesn't (e.g., HIE), it maybe just isn't calculating it. 

It may be worth some consideration to change the way "is_protein" attribute is defined to encompass "engineered" residues as well, with non-standard three letter names, and also to make sure dihedrals are calculated for this wider set of defined residues. A workaround that @cxhernandez nicely suggested in the meantime is to just rename those engineered residues to standard names for now, but it would be nice to ultimately fix these definitions in the code base. Maybe we can discuss sensible ways to make "is_protein" definition more flexible without having too wide of a scope?

Cheers and thanks as always for your thoughts,
Evan

**Comment by cxhernandez** [comment]
"HID" and "HIE" are fairly common (dunno about "ASH"), so maybe we can just add those residue codes to `_PROTEIN_RESIDUES ` in [`residue_names.py`](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/core/residue_names.py). 

One option could be to pass non-standard residue codes into `md.load` like:

``` python
traj = md.load('traj.lh5', include_protein=['HIE', 'HIS'], include_solvent=['DMSO'])
```

But I don't know if it's a good idea to change the API just to support anybody's crazy three-letter code convention since it could the code-base more unwieldily. It's probably just worth noting in the docs that this pre-defined list exists and is easily editable.

What do you think @rmcgibbo?

**Comment by evanfeinberg** [comment]
Agreed.

I added HIE, HIS, HIP, and ASH to "/home/enf/anaconda/pkgs/mdtraj-1.3.0-np19py27_0/lib/python2.7/site-packages/mdtraj-1.3.0-py2.7-linux-x86_64.egg/mdtraj/core/residue_names.py" and to "/home/enf/anaconda/lib/python2.7/site-packages/mdtraj-1.3.0-py2.7-linux-x86_64.egg/mdtraj/core/residue_names.py", but it still doesn't seem to be computing any dihedral angles for those histidines. Are there any other places where I can just edit the code such that mdtraj will compute dihedrals for those residues?

**Comment by evanfeinberg** [comment]
Alternatively: is there a way to rename residues on-the-fly in mdtraj, so I can map all histidines to HIE, map all aspartates to Asp, etc., without editing the PDB File itself?

**Comment by cxhernandez** [comment]
Are the atom names non-standard as well? This is the only hard-coded thing in `dihedral.py` that I could find were [lines 275-298](https://github.com/mdtraj/mdtraj/blob/master/mdtraj/geometry/dihedral.py#L275-298).

**Comment by evanfeinberg** [comment]
The atom names looked all standard, i.e. CA, CD1, etc. Did I edit the wrong residue_names.py version?

**Comment by cxhernandez** [comment]
You should be able to assign residue info to the topology object:

``` python
traj.topology.residue(i).is_protein = True
```

**Comment by cxhernandez** [comment]
It seems like you did with `/home/enf/anaconda/lib/python2.7/site-packages/mdtraj-1.3.0-py2.7-linux-x86_64.egg/mdtraj/core/residue_names.py`, but the sure way is to `pip uninstall mdtraj` and then compile mdtraj from source (after adding the missing residue names).

**Comment by evanfeinberg** [comment]
OK that sounds good, going to try that now, thanks!!

**Comment by evanfeinberg** [comment]
OK so I initially installed from omnia: conda install -c https://conda.binstar.org/omnia mdtraj
and now I'm uninstalling with "conda remove mdtraj" but it's stuck at stage "Fetching package metadata:" and has not progressed. Has anyone run into this issue before?

**Comment by cxhernandez** [comment]
try using pip

**Comment by evanfeinberg** [comment]
already did :/

pip uninstall mdtraj
Cannot remove entries from nonexistent file /home/enf/anaconda/lib/python2.7/site-packages/easy-install.pth

**Comment by cxhernandez** [comment]
Try including the channel (`-c`) flag

**Comment by evanfeinberg** [comment]
--offline ended up working. probably some weird sherlock issue, as per usual

**Comment by evanfeinberg** [comment]
No joy in Mudville. I uninstalled both MDtraj and MSMBuilder. Then I DL'd the newest version of MDtraj from github, changed residue_names.py to include the new residue names, then installed MDtraj from source then installed MSMBuilder from source. All instances of residue_names.py in my ~/ directory have the new residue names. Yet, when I try to use DihedralFeaturizer, I get different numbers of features for different trajectories, even though the only chemical difference is proton locations.

**Comment by mpharrigan** [comment]
Just a quick note. The proper way to fix this is to add "aliases" for the "non-standard" residue names in the pdb parser. Per @rmcgibbo , the internal representation is supposed to be canonicalized 

cc #637

**Comment by mpharrigan** [comment]
Also, @cxhernandez I don't think you can set `is_protein` like that. It is a read only `property`

**Comment by evanfeinberg** [comment]
Update: so, on my last post in the msmvuilder thread (sorry I posted it in the wrong place), I determined that the is_protein issue is an adjunct to another problem: in the topology file, a ton of residues are split up into separate objects! This is probably because, even though they have the same residue number, chain, and name in the pdb file, the likes for a single residue are split across the file. Take a look at HIE93 in the system.pdb file. This is a result of it being unordered in the maestro file. I have no idea why the authors of that paper decided to format it that way. Anyway, I dug through the code and I think we can create a relatively easy workaround by modifying the _atom_sequence function in the dihedral.py file. Please see my other post in the msmbuilder page for details. There's more than one way to fix this, but I think modifying _atom_sequence might be the easiest for now.

**Comment by evanfeinberg** [comment]
The PDB file that MDTraj is reading in perhaps imperfectly can be found at: 
/home/harrigan/compute/wetmsm/gpcr/des/system_mae_to_pdb/des_trajs/DESRES-Trajectory_pnas2011b-B-01-all/system.pdb

TL;DR: it would be helpful if either (1) PDB reading were modified such that all atoms of same res #, chain, and name were placed in the same residue object, regardless of where their lines physically are in the pdb file, or (2) _atom_sequence() in dihedral.py were made more flexible to circumvent this issue.

**Comment by evanfeinberg** [comment]
I'm going to attempt a workaround now by writing a Python script to "fix" PDB files that have this issue, I'lll report back if it works...

**Comment by evanfeinberg** [comment]
Alright so I think this did the trick. I need to test it more, so consider it in alpha:


```
'''
Many PDB Files have an issue where a single chemical residue (e.g., a single histidine) will have each of its atoms, while
	chemically bonded, spread out across the PDB file. This confuses many PDB readers, as it will assume each non-contiguous
	line for that residue corresponds to separate residues. This class has a function that takes as inputs the name of the original
	pdb file and the name of a new file to which the user would like to write. It then finds all residues corresponding to proteins,
	and stores them in a dict, with keys as a tuple of (residue_name, residue_chain, and residue_number) and with values as a 
	list of physical lines in the PDB file that correspond to that unique residue identifier. The function then prints the initial header,
	prints the exact same PDB file lines just in a different order that maintains contiguousness within each residue, and then prints
	all membrane, solvent, and other lines.


'''
class PDB_Order_Fixer:

	def __init__(self, filename, new_filename):
		self.filename= filename
		self.new_filename = new_filename

	def fix_pdb(self):
		filename = self.filename
		new_filename = self.new_filename

		residue_lines = {}
		residue_order = []
		non_protein_lines = []


		pdb_file = open(filename, "rb")
		lines = pdb_file.readlines()
		i = 0

		while lines[i][0:4] != "ATOM" and lines[i][0:6] != "HETATM":
			residue_key = (i, 0, 0)
			residue_order.append(residue_key)
			residue_lines[residue_key] = [i]
			i += 1

		print("Found first atom line, line %d" %i)

		while i < (len(lines) - 1):
			line = lines[i].split()

			if len(line) < 2:
				new_protein_lines.append(i)
				i += 1
				continue

			res_name = line[3]

			if res_name[0:3] in ["POP", "TIP", "HOH", "MEM"]:
				non_protein_lines.append(i)
				i += 1
				continue

			chain = line[4]
			res_number = line[5]
			residue_key = (res_name, chain, res_number)

			if residue_key in residue_lines.keys():
				residue_lines[residue_key].append(i)
			else:
				residue_lines[residue_key] = [i]
				residue_order.append(residue_key)

			i += 1

		print("read a total of %d lines" %i)
		print("finished reading file")

		new_file = open(new_filename, "wb")

		for residue_key in residue_order:
			for line_number in residue_lines[residue_key]:
				new_file.write(lines[line_number])

		for line_number in non_protein_lines:
			new_file.write(lines[line_number])

		print("finished writing new file")

		new_file.close()

```

**Comment by evanfeinberg** [comment]
Alright @mpharrigan , @cxhernandez , @rmcgibbo , thank you all for your advice. I have two updates. (1) A successful workaround to the issue. (2) An unsuccessful solution that should nonetheless work. 

(1) I wrote a custom dihedral featurizer that includes its own "topology fixing" function, 4-tuple atom sequencing functions for each dihedral type (only 3 right now). The code is available here:

https://gist.github.com/evanfeinberg/b6f9c8bc6c54f98d09dc#file-custom_featurizer

 This solves the issue by fixing a *copy* of the topology file, so it doesn't mess up atom coordinates of the actual trajectory, using the fixed top file to generate dihedral atom sequences, and then calling compute_dihedral to find the dihedral (and then sin/cos). This seems to recover all the requisite dihedral angles. On limited benchmarks it works as fast as the standard msmbuilder DihedralFeaturizer class if anyone wants to try it out.

(2) Two methods failed, but proved useful for other things later on. 

(a) reordering PDB files, now available in this gist: 

/evanfeinberg/1db727d37ba27d5c59dd#file-pdb_order_fixer-py

However, it does not work as a topology file for DCD trajectories, as it messes up ordering. Only use this if you are just going to load the resultant PDB itself into another program for manipulation or analysis. It worked wonders, for example, for fixing PDB's before loading into PyMol and Maestro.

(b) Fixing the topology itself. When MDTraj reads in a PDB topology and DCD file, if the residues are split into fragments across the DCD file, it then creates a *separate* residue object for several fragments, even though they are actually part of the same residue. I tried several ways of "fixing" the MDTraj.trajectory.topology object of a given trajectory, i.e., placing all atoms part of the same chemical residue into the same residue object, and changing the residues to which those atoms refer. See gist here:

https://gist.github.com/evanfeinberg/e7d3ab9e15505e872fe9

Unfortunately, this completely messes up the geometry of the protein. Somehow, even though everything else is the same, the xyz coordinates to which each atom refers is messed up. Any ideas how to fix this? I think this topology fixing is ultimately the best solution to this problem, even though (1) is a good workaround (and generally useful, to boot) for now.
